tokendito/duo_helpers.py

```diff
@@ -0,0 +1,363 @@
+# vim: set filetype=python ts=4 sw=4
+# -*- coding: utf-8 -*-
+"""This module handles Duo operations."""
+from __future__ import (absolute_import, division,
+                        print_function, unicode_literals)
+
+from builtins import (ascii, bytes, chr, dict, filter, hex, input,  # noqa: F401
+                      int, list, map, next, object, oct, open, pow, range,
+                      round, str, super, zip)
+import json
+import logging
+import sys
+import time
+from urllib.parse import unquote, urlparse
+
+from bs4 import BeautifulSoup
+from future import standard_library
+import requests
+from tokendito import helpers
+from tokendito import settings
+standard_library.install_aliases()
+
+
+def prepare_duo_info(selected_okta_factor):
+    """Aggregate most of the parameters needed throughout the Duo authentication process.
+
+    :param selected_okta_factor: dict response describing Duo factor in Okta.
+    :return duo_info: dict of parameters for Duo
+    """
+    duo_info = {}
+    okta_factor = selected_okta_factor["_embedded"]["factor"]["_embedded"]["verification"]
+    duo_info["okta_factor"] = okta_factor
+    duo_info["factor_id"] = selected_okta_factor["_embedded"]["factor"]["id"]
+
+    duo_info["state_token"] = selected_okta_factor["stateToken"]
+    duo_info["okta_callback_url"] = okta_factor["_links"]["complete"]["href"]
+    duo_info["tx"] = okta_factor["signature"].split(":")[0]
+    duo_info["app_sig"] = okta_factor["signature"].split(":")[1]
+    duo_info["parent"] = "{}/signin/verify/duo/web".format(
+        settings.okta_org)
+    duo_info["host"] = okta_factor["host"]
+    duo_info["sid"] = ""
+
+    version = okta_factor["_links"]["script"]["href"].split("-v")[1]
+    duo_info["version"] = version.strip(".js")
+
+    return duo_info
+
+
+def duo_api_post(url, params={}, headers={}, payload={}):
+    """Error handling and response parsing wrapper for Duo POSTs.
+
+    :param url: The URL being connected to.
+    :param params: URL query parameters.
+    :param headers: Request headers.
+    :param payload: Request body.
+    :return response: Response to the API request.
+    """
+    try:
+        response = requests.request(
+            'POST', url, params=params, headers=headers, data=payload)
+    except Exception as request_issue:
+        logging.error(
+            "There was an error connecting to the Duo API: \n{}".format(request_issue))
+        sys.exit(1)
+
+    json_message = None
+    try:
+        json_message = response.json()
+    except ValueError:
+        logging.debug("Non-json response from Duo API: \n{}".format(response))
+
+    if response.status_code != 200:
+        logging.critical("Your Duo authentication has failed with status {}.".format(
+            response.status_code))
+        if json_message and json_message["stat"].lower() != "ok":
+            logging.critical("\n{}".format(
```

Comment:
logging.critical("\n{}, {}".format(response.status_code, json_message["message"])) 
instead of 
logging.critical("\n{}".format(response.status_code, json_message["message"]))